src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -563,6 +568,73 @@ private static IEnumerable<string> DiscoverBindingTypes(IEnumerable<FunctionMeta
             return bindingTypes;
         }
 
+        // Validate that for any precompiled assembly, all functions have the same configuration precedence.
+        // Also load any extensions from the precompiled directory.
+        private void VerifyPrecompiled(Collection<FunctionMetadata> functions)
+        {
+            HashSet<string> extensionFolders = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            Dictionary<string, bool> mapAssemblySettings = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
+            foreach (var function in functions)
+            {
+                var scriptFile = function.ScriptFile;
+                if (scriptFile.EndsWith(".dll", StringComparison.OrdinalIgnoreCase))
+                {
+                    bool isDirect = function.IsDirect;
+                    bool prevIsDirect;
+                    if (mapAssemblySettings.TryGetValue(scriptFile, out prevIsDirect))
+                    {
+                        if (prevIsDirect != isDirect)
+                        {
+                            // Error. All entries pointing to the same dll must have the same value for IsDirect
+                            string msg = string.Format(CultureInfo.InvariantCulture, "Configuration error: All functions in {0} must have the same config precedence,",
+                                scriptFile);
+                            TraceWriter.Info(msg);
+                            _startupLogger?.LogInformation(msg);
+
+                            // Hard failure
+                            functions.Clear();
+                            return;
+                        }
+                    }
+                    mapAssemblySettings[scriptFile] = isDirect;
+
+                    string path = Path.GetDirectoryName(scriptFile);
+                    extensionFolders.Add(path);
+                }
+            }
+
+            // For any direct loaded assemblies, check for extensions.
+            foreach (var path in extensionFolders)
+            {
+                LoadCustomExtensions(path);
```

Comment:
This gives us exactly the customer experience we want. If a user references an extension in their VS project, this will cause it to get loaded in the runtime.  

---
In reply to: [129920865](https://github.com/Azure/azure-webjobs-sdk-script/pull/1717#discussion_r129920865) [](ancestors = 129920865)